websocket_client.py

```diff
@@ -0,0 +1,463 @@
+                    self.emit("output", self._pending_output, msg["final"])
+                if msg["final"]:
+                    self._pending_output = ""
+            case "client_tool_invocation":
+                await self._handle_client_tool_call(
+                    msg["toolName"], msg["invocationId"], msg["parameters"]
+                )
+            case "debug":
+                logging.info(f"debug: {msg['message']}")
+            case _:
+                logging.warning(f"Unhandled message type: {msg['type']}")
+
+    async def _handle_client_tool_call(
+        self, tool_name: str, invocation_id: str, parameters: dict[str, Any]
+    ):
+        logging.info(f"client tool call: {tool_name}")
+        if tool_name == "getSecretMenu":
+            menu = [
+                {
+                    "date": datetime.date.today().isoformat(),
+                    "items": [
+                        {
+                            "name": "Banana Smoothie",
+                            "price": "$4.99",
+                        },
+                        {
+                            "name": "Butter Pecan Ice Cream (one scoop)",
+                            "price": "$2.99",
+                        },
+                    ],
+                },
+            ]
+            result = {
+                "result": json.dumps(menu),
+            }
+        else:
+            result = {
+                "errorType": "undefined",
+                "errorMessage": f"Unknown tool: {tool_name}",
+            }
+        result["invocationId"] = invocation_id
+        result["type"] = "client_tool_result"
+        await self._socket.send(json.dumps(result))
+
+    async def _pump_audio(self, source: LocalAudioSource):
+        async for chunk in source.stream():
+            if self._socket is None:
+                continue
+            await self._socket.send(chunk)
+
+
+async def _async_close(*awaitables_or_none: Awaitable | None):
+    coros = [coro for coro in awaitables_or_none if coro is not None]
+    if coros:
+        maybe_exceptions = await asyncio.shield(
+            asyncio.gather(*coros, return_exceptions=True)
+        )
+        non_cancelled_exceptions = [
+            exc
+            for exc in maybe_exceptions
+            if isinstance(exc, Exception)
+            and not isinstance(exc, asyncio.CancelledError)
+        ]
+        if non_cancelled_exceptions:
+            to_report = (
+                non_cancelled_exceptions[0]
+                if len(non_cancelled_exceptions) == 1
+                else ExceptionGroup("Multiple failures", non_cancelled_exceptions)
+            )
+            logging.warning("Error during _async_close", exc_info=to_report)
+
+
+async def _async_cancel(*tasks_or_none: asyncio.Task | None):
+    tasks = [task for task in tasks_or_none if task is not None and task.cancel()]
+    await _async_close(*tasks)
+
+
+async def _get_join_url() -> str:
```

Comment:
"""
    Create a new Ultravox call and returns the join URL.
    
    Configures the session based on command-line arguments including:
    - System prompt
    - Voice selection
    - Temperature
    - Special features (secret menu, experimental messages)
    
    Returns:
        str: Ultravox join URL for joining the voice session
        
    Raises:
        ValueError: If ULTRAVOX_API_KEY is not set
        aiohttp.ClientError: For API request failures
    """